### PR TITLE
[SPARK-21012][Submit] Add glob support for resources adding to Spark

### PR DESCRIPTION
**Pull request by jerryshao**
Current "--jars (spark.jars)", "--files (spark.files)", "--py-files (spark.submit.pyFiles)" and "--archives (spark.yarn.dist.archives)" only support non-glob path. This is OK for most of the cases, but when user requires to add more jars, files into Spark, it is too verbose to list one by one. So here propose to add glob path support for resources.

Also improving the code of downloading resources.

## How was this patch tested?

UT added, also verified manually in local cluster.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 1]
```diff
@@ -858,19 +844,33 @@ object SparkSubmit extends CommandLineUtils {
     require(path != null, "path cannot be null.")
     val uri = Utils.resolveURI(path)
     uri.getScheme match {
-      case "file" | "local" =>
-        path
-
+      case "file" | "local" => path
+      case "https" | "http" | "ftp" => path
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: This is wrong, right? These need to be downloaded, and now they won't be.

(Whether the `FileSystem` API can download those is a separate question... if it doesn't, you'll end up with another implementation of `Utils.doFetchFile`.)

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 2]
```diff
@@ -858,19 +844,33 @@ object SparkSubmit extends CommandLineUtils {
...
         printStream.println(s"Downloading ${uri.toString} to ${tmpFile.getAbsolutePath}.")
         // scalastyle:on println
         fs.copyToLocalFile(new Path(uri), new Path(tmpFile.getAbsolutePath))
         Utils.resolveURI(tmpFile.getAbsolutePath).toString
     }
   }
+
+  private def resolveGlobPaths(paths: String, hadoopConf: HadoopConfiguration): String = {
+    require(paths != null, "paths cannot be null.")
+    paths.split(",").filter(_.trim.nonEmpty).flatMap { path =>
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: minor, but you should actually trim the strings, not just for filtering purposes.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala (hunk 1 above)
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: My original thinking is that `FileSystem` API doesn't support such API, so I treat them separately without downloading. Will verify it.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 3]
```diff
@@ -310,33 +310,28 @@ object SparkSubmit extends CommandLineUtils {
...
-    // In client mode, download remote files.
-    if (deployMode == CLIENT) {
-      val hadoopConf = new HadoopConfiguration()
-      args.primaryResource = Option(args.primaryResource).map(downloadFile(_, hadoopConf)).orNull
-      args.jars = Option(args.jars).map(downloadFileList(_, hadoopConf)).orNull
-      args.pyFiles = Option(args.pyFiles).map(downloadFileList(_, hadoopConf)).orNull
-      args.files = Option(args.files).map(downloadFileList(_, hadoopConf)).orNull
-    }
+    val hadoopConf = new HadoopConfiguration()
+    val targetDir = Files.createTempDirectory("tmp").toFile
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: I see this comes from the previous code, but who's deleting this directory?

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 4]
```diff
@@ -310,33 +310,28 @@ object SparkSubmit extends CommandLineUtils {
...
+    val sparkConf = new SparkConf(loadDefaults = false).setAll(args.sparkProperties)
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: Hmm... I think there might be a problem with logging when calling these classes here. This code is running before the app's main class is called, and it may initialize the logging system (e.g. `Utils.doFetchFile` later calls logging methods in all code paths, initializing the logging system, and so does `SecurityManager`).

This will be a problem, for example, for the spark-shell, which triggers a slightly different code path when initializing logging. Because now the logging system will be initialized by a class that is not the REPL's `Main` class, that will be lost.

Long story short, with the code you're adding, the shell should still be initialized with `WARN` as the default log level. Can you run `spark-shell` with a remote jar in the jars list and verify it still maintains that?

**Review by vanzin — COMMENTED** [review]
I have one question that is more of a question for the old code that introduced the download feature, and one concern about the code being added.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 5]
```diff
@@ -845,32 +840,54 @@ object SparkSubmit extends CommandLineUtils {
    */
   private[deploy] def downloadFileList(
       fileList: String,
+      targetDir: File,
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: Let's explain the meaning of each param.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 6]
```diff
@@ -845,32 +840,54 @@ object SparkSubmit extends CommandLineUtils {
...
   }
 
   /**
    * Download a file from the remote to a local temporary directory. If the input path points to
    * a local path, returns it with no operation.
    */
-  private[deploy] def downloadFile(path: String, hadoopConf: HadoopConfiguration): String = {
+  private[deploy] def downloadFile(
+      path: String,
+      targetDir: File,
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: ditto

File: core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala [hunk 7]
```diff
@@ -752,25 +793,35 @@ class SparkSubmitSuite
 
   test("downloadFile - invalid url") {
     intercept[IOException] {
-      SparkSubmit.downloadFile("abc:/my/file", new Configuration())
+      val sparkConf = new SparkConf()
```
Comment (jiangxb1987) [inline core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala]: Should we create a `testDownloadFile()` function and merge the duplicated test code?

**Review by jiangxb1987 — COMMENTED** [review]
Looks good overall, only some minor comments.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala (hunk 3 above)
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: From my understanding currently no code is responsible for deleting, let me check the code.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala (hunk 4 above)
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: @vanzin yes, your concern is correct. `Logging` module will be initialized by other class, thus default log level constraint is broken.

The purpose here is to use `Utils.doFetchFile` without duplication. One solution I could think of is to add the similar code here to avoid logging initialization.

**Comment by jiangxb1987** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 8]
```diff
@@ -841,36 +845,132 @@ object SparkSubmit extends CommandLineUtils {
...
     val uri = Utils.resolveURI(path)
     uri.getScheme match {
-      case "file" | "local" =>
-        path
+      case "file" | "local" => path
+      case "http" | "https" | "ftp" =>
+        val uc = uri.toURL.openConnection()
+        uc match {
+          case https: HttpsURLConnection =>
+            val trustStore = sparkProperties.get("spark.ssl.fs.trustStore")
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: Should we make this a common util?

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 9]
```diff
@@ -841,36 +845,132 @@ object SparkSubmit extends CommandLineUtils {
...
+              .orElse(sparkProperties.get("spark.ssl.trustStore"))
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: Should we move these properties to internal/config?

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala (hunk 8 above)
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: Also, should we test against this?

Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: There is a common util method in `Utils`, but the thing is that using that method as I did before will initialize `Logging` prematurely, that will have some issues as mentioned by @vanzin , so that's why I reimplemented here.

Do you have any thought?

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala (hunk 9 above)
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: These configurations are dynamic configurations based on the component `spark.ssl.${ns}.xxx`, It is not easy to leverage `internal/config`. Let me think about it.

Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: make sense, let's keep this way.

**Comment by jiangxb1987** [comment]
LGTM, cc @cloud-fan

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala [hunk 10]
```diff
@@ -520,7 +520,7 @@ private[deploy] class SparkSubmitArguments(args: Seq[String], env: Map[String, S
         |                              (Default: client).
         |  --class CLASS_NAME          Your application's main class (for Java / Scala apps).
         |  --name NAME                 A name of your application.
-        |  --jars JARS                 Comma-separated list of local jars to include on the driver
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: This is a new feature(support non-local jars), shall we create a separated jira ticket?

Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: and we should also add a test case for it.

Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: Thanks @cloud-fan for your review. Supporting remote jar has been addressed in #18078 , but it still left some codes and docs should be updated, also it doesn't support downloading jars from HTTP(S), FTP, here also adding this support.

So basically this PR address the left problem of #18078 and add glob support.

**Comment by cloud-fan** [comment]
LGTM, merging to master!
